src/compas/geometry/curves/nurbs.py

```diff
@@ -0,0 +1,462 @@
+
+    def copy(self):
+        """Make an independent copy of the current curve."""
+        return NurbsCurve.from_parameters(
+            self.points,
+            self.weights,
+            self.knots,
+            self.multiplicities,
+            self.degree,
+            self.is_periodic
+        )
+
+    def transform(self, T):
+        """Transform this curve."""
+        raise NotImplementedError
+
+    def transformed(self, T):
+        """Transform a copy of the curve."""
+        copy = self.copy()
+        copy.transform(T)
+        return copy
+
+    def space(self, n=10):
+        """Compute evenly spaced parameters over the curve domain."""
+        u, v = self.domain
+        return linspace(u, v, n)
+
+    def xyz(self, n=10):
+        """Compute point locations corresponding to evenly spaced parameters over the curve domain."""
+        return [self.point_at(param) for param in self.space(n)]
+
+    def locus(self, resolution=100):
+        """Compute the locus of all points on the curve.
+
+        Parameters
+        ----------
+        resolution : int
+            The number of intervals at which a point on the
+            curve should be computed. Defaults to 100.
+
+        Returns
+        -------
+        list
+            Points along the curve.
+        """
+        return self.xyz(resolution)
+
+    def point_at(self, u):
+        """Compute a point on the curve.
+
+        Parameters
+        ----------
+        u : float
+            The value of the curve parameter. Must be between 0 and 1.
+
+        Returns
+        -------
+        Point
+            the corresponding point on the curve.
+        """
+        raise NotImplementedError
+
+    def tangent_at(self, u):
+        """Compute the tangent vector at a point on the curve.
+
+        Parameters
+        ----------
+        u : float
+            The value of the curve parameter. Must be between 0 and 1.
+
+        Returns
+        -------
+        Vector
+            The corresponding tangent vector.
+
+        """
+        pass
+
+    def curvature_at(self, u):
+        """Compute the curvature at a point on the curve.
+
+        Parameters
+        ----------
+        u : float
+            The value of the curve parameter. Must be between 0 and 1.
+
+        Returns
+        -------
+        Vector
+            The corresponding curvature vector.
+
+        """
+        raise NotImplementedError
+
+    def frame_at(self, u):
+        """Compute the local frame at a point on the curve.
+
+        Parameters
+        ----------
+        u : float
+            The value of the curve parameter. Must be between 0 and 1.
+
+        Returns
+        -------
+        Frame
+            The corresponding local frame.
+
+        """
+        raise NotImplementedError
+
+    def closest_point(self, point, distance=None):
+        """Compute the closest point on the curve to a given point."""
+        raise NotImplementedError
```

Comment:
The idea is that these method's implementation are coming from OCC or Rhino SDK, right? If so, two questions:
1) how about throwing a plugin-specific type of exception, e.g the existing `compas.plugins.PluginNotInstalledError` or maybe a new, more precise one, like `compas.plugins.NoPluggableImplFound`
2) when the surfaces are *not* based on the OCC implementation, how exactly are these things bound to the Rhino SDK ones? I think we discussed this, but I'm lost about where the link is.... :P

Reply:
i think the idea would be to intercept the object instantiation like with the artists and have both occ and Rhino provide a plugin for that, but not sure

Reply:
got it! yep, that's what I thought we discussed, but was a bit lost about where we were on that

Reply:
this is now solved for the OCC implementation. not yet for Rhino